### PR TITLE
EN-10523: Web server refactoring

### PR DESCRIPTION
**Pull request by bogdan-rosianu**
What has been done:
- refactored API routes handler: from exported, state-less functions to methods of structures;
- eliminate the `facade` that was set in gin's context -> instead, have it as a member in structures;
- copied and adapted all the endpoints and their unit tests.

What's left to do:
- integrate the new structures inside API starter;
- remove old routes handlers;
- integrate in nodeRunner the new version -> the gin server shouldn't be restarted anymore.

### CODE REVIEW

File: api/groups/vmValuesGroup_test.go [hunk 1]
```diff
@@ -0,0 +1,355 @@
...
+type vmOutputResponse struct {
+	Data  *vmcommon.VMOutput `json:"data"`
+	Error string             `json:"error"`
+}
+
+func init() {
+	gin.SetMode(gin.TestMode)
+}
+
+const DummyScAddress = "00000000000000000500fabd9501b7e5353de57a4e319857c2fb99089770720a"
```
Comment (iulianpascalau) [inline api/groups/vmValuesGroup_test.go:62]: comment?

File: api/groups/blockGroup.go [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+func (bg *blockGroup) getFacade() blockFacadeHandler {
+	bg.mutFacade.RLock()
+	defer bg.mutFacade.RUnlock()
+
+	return bg.facade
+}
+
+// UpdateFacade will update the facade
+func (bg *blockGroup) UpdateFacade(newFacade interface{}) error {
+	if newFacade == nil {
```
Comment (iulianpascalau) [inline api/groups/blockGroup.go:161]: unit tests on all untested UpdateFacade functions.

**Review by iulianpascalau — CHANGES_REQUESTED** [review]
Minor stuff :+1: . Compared with the existing api handlers

File: api/groups/addressGroup.go [hunk 3]
```diff
@@ -0,0 +1,707 @@
...
+	Nonce           uint64   `json:"nonce,omitempty"`
+	Creator         string   `json:"creator,omitempty"`
+	Royalties       string   `json:"royalties,omitempty"`
+	Hash            []byte   `json:"hash,omitempty"`
+	URIs            [][]byte `json:"uris,omitempty"`
+	Attributes      []byte   `json:"attributes,omitempty"`
+}
+
+// NewAddressGroup returns a new instance of addressGroup
+func NewAddressGroup(facadeHandler interface{}) (*addressGroup, error) {
```
[the remark this answers is not in the document]
Comment (bogdan-rosianu) [reply api/groups/addressGroup.go:71]: done

File: api/groups/addressGroup.go [hunk 4]
```diff
@@ -0,0 +1,707 @@
...
+				Error: fmt.Sprintf("%s: %s", errors.ErrGetESDTTokens.Error(), err.Error()),
+				Code:  shared.ReturnCodeInternalError,
+			},
+		)
+		return
+	}
+
+	formattedTokens := make(map[string]*esdtNFTTokenData)
+	for tokenID, esdtData := range tokens {
+		tokenData := &esdtNFTTokenData{
```
[the remark this answers is not in the document]
Comment (bogdan-rosianu) [reply api/groups/addressGroup.go:657]: done

File: api/groups/addressGroup_test.go [hunk 5]
```diff
@@ -0,0 +1,990 @@
...
+	apiErrors "github.com/ElrondNetwork/elrond-go/api/errors"
+	"github.com/ElrondNetwork/elrond-go/api/groups"
+	"github.com/ElrondNetwork/elrond-go/api/mock"
+	"github.com/ElrondNetwork/elrond-go/api/shared"
+	"github.com/ElrondNetwork/elrond-go/config"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+type AccountResponse struct {
```
[the remark this answers is not in the document]
Comment (bogdan-rosianu) [reply api/groups/addressGroup_test.go:24]: done

File: api/groups/baseGroup.go [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+		properties := getEndpointProperties(ws, handlerData.Path, apiConfig)
+
+		if !properties.isOpen {
+			log.Debug("endpoint is closed", "path", handlerData.Path)
+			continue
+		}
+
+		middlewares := make([]gin.HandlerFunc, 0)
+		for _, middleware := range additionalMiddlewares {
+			if middleware == nil ||
```
[the remark this answers is not in the document]
Comment (bogdan-rosianu) [reply api/groups/baseGroup.go:43]: removed the additionalMiddlewares section. I originally thought that I'll use them to inject the source/global/logging middlewares, but they are set directly into parent gin Router

File: api/groups/baseGroup.go [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+				middleware.MiddlewareHandlerFunc() == nil ||
+				middleware.IsInterfaceNil() {
+				continue
+			}
+			middlewares = append(middlewares, middleware.MiddlewareHandlerFunc())
+		}
+
+		beforeSpecifiesMiddlewares, afterSpecificMiddlewares := extractSpecificMiddlewares(handlerData.AdditionalMiddlewares)
+		for _, specificMiddleware := range beforeSpecifiesMiddlewares {
```
[the remark this answers is not in the document]
Comment (bogdan-rosianu) [reply api/groups/baseGroup.go:52]: done. also confirmed within a unit test that the code won't panic in case that the slice to append is nil

File: api/groups/baseGroup.go [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+	for _, handlerData := range bg.endpoints {
+		properties := getEndpointProperties(ws, handlerData.Path, apiConfig)
+
+		if !properties.isOpen {
+			log.Debug("endpoint is closed", "path", handlerData.Path)
+			continue
+		}
+
+		middlewares := make([]gin.HandlerFunc, 0)
+		for _, middleware := range additionalMiddlewares {
```
[the remark this answers is not in the document]
Comment (bogdan-rosianu) [reply api/groups/baseGroup.go:42]: no need to. after the refactoring suggested above, the `for` block is much simpler

File: api/groups/baseGroup.go [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+			}
+			middlewares = append(middlewares, middleware.MiddlewareHandlerFunc())
+		}
+
+		beforeSpecifiesMiddlewares, afterSpecificMiddlewares := extractSpecificMiddlewares(handlerData.AdditionalMiddlewares)
+		for _, specificMiddleware := range beforeSpecifiesMiddlewares {
+			middlewares = append(middlewares, specificMiddleware)
+		}
+		middlewares = append(middlewares, handlerData.Handler)
+		for _, specificMiddleware := range afterSpecificMiddlewares {
```
[the remark this answers is not in the document]
Comment (bogdan-rosianu) [reply api/groups/baseGroup.go:56]: done

File: api/groups/baseGroup.go [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+		}
+	}
+
+	return endpointProperties{
+		isOpen: false,
+	}
+}
+
+// IsInterfaceNil returns true if there is no value under the interface
+func (bg *baseGroup) IsInterfaceNil() bool {
```
[the remark this answers is not in the document]
Comment (bogdan-rosianu) [reply api/groups/baseGroup.go:110]: interesting find. moved the IsInterfaceNil to all gorups

File: api/groups/export_test.go [hunk 11]
```diff
@@ -0,0 +1,11 @@
+package groups
+
+import "github.com/ElrondNetwork/elrond-go/process"
+
+const ExecManualTrigger = execManualTrigger
```
[the remark this answers is not in the document]
Comment (bogdan-rosianu) [reply api/groups/export_test.go:5]: added

File: api/groups/hardforkGroup.go [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+	if newFacade == nil {
+		return errors.ErrNilFacadeHandler
+	}
+	castedFacade, ok := newFacade.(hardforkFacadeHandler)
+	if !ok {
+		return errors.ErrFacadeWrongTypeAssertion
+	}
+
+	hg.mutFacade.Lock()
+	hg.facade = castedFacade
```
[the remark this answers is not in the document]
Comment (bogdan-rosianu) [reply api/groups/hardforkGroup.go:127]: done

File: api/groups/nodeGroup_test.go [hunk 13]
```diff
@@ -0,0 +1,406 @@
...
+	})
+
+	t.Run("should work", func(t *testing.T) {
+		hg, err := groups.NewNodeGroup(&mock.Facade{})
+		require.NoError(t, err)
+		require.NotNil(t, hg)
+	})
+}
+
+type GeneralResponse struct {
```
[the remark this answers is not in the document]
Comment (bogdan-rosianu) [reply api/groups/nodeGroup_test.go:56]: made them unexported. Only the fields need to be exported because they are meant to map JSON fields

File: api/groups/proofGroup.go [hunk 14]
```diff
@@ -0,0 +1,286 @@
...
+	"github.com/ElrondNetwork/elrond-go/api/middleware"
+	"github.com/ElrondNetwork/elrond-go/api/shared"
+	"github.com/gin-gonic/gin"
+)
+
+const (
+	getProofCurrentRootHashEndpoint = "/proof/address/:address"
+	getProofEndpoint                = "/proof/root-hash/:roothash/address/:address"
+	verifyProofEndpoint             = "/proof/verify"
+
```
[the remark this answers is not in the document]
Comment (bogdan-rosianu) [reply api/groups/proofGroup.go:20]: done

File: api/shared/shared.go [hunk 15]
```diff
@@ -6,6 +6,20 @@ import (
 	"github.com/gin-gonic/gin"
 )
 
+// AdditionalMiddleware holds the data needed for adding a middleware to an API endpoint
+type AdditionalMiddleware struct {
+	Middleware gin.HandlerFunc
+	Before bool
```
[the remark this answers is not in the document]
Comment (bogdan-rosianu) [reply api/shared/shared.go:12]: good stuff. refactored

File: api/groups/transactionGroup.go [hunk 16]
```diff
@@ -0,0 +1,578 @@
...
+	ComputeTransactionGasLimit(tx *transaction.Transaction) (*transaction.CostResponse, error)
+	EncodeAddressPubkey(pk []byte) (string, error)
+	GetThrottlerForEndpoint(endpoint string) (core.Throttler, bool)
+	IsInterfaceNil() bool
+}
+
+type transactionGroup struct {
+	facade    transactionFacadeHandler
+	mutFacade sync.RWMutex
+	*baseGroup
```
[the remark this answers is not in the document]
Comment (bogdan-rosianu) [reply api/groups/transactionGroup.go:53]: done

File: api/groups/vmValuesGroup_test.go (hunk 1 above)
Comment (bogdan-rosianu) [reply api/groups/vmValuesGroup_test.go:62]: no nned. made it unexpeorted

File: api/groups/blockGroup.go (hunk 2 above)
Comment (bogdan-rosianu) [reply api/groups/blockGroup.go:161]: I would leave this for the next PR(s) as there are also other components that need to be unit tested

File: api/groups/addressGroup.go [hunk 17]
```diff
@@ -0,0 +1,707 @@
...
+	GetESDTsWithRole(address string, role string) ([]string, error)
+	GetAllESDTTokens(address string) (map[string]*esdt.ESDigitalToken, error)
+	GetKeyValuePairs(address string) (map[string]string, error)
+	IsInterfaceNil() bool
+}
+
+type addressGroup struct {
+	facade    addressFacadeHandler
+	mutFacade sync.RWMutex
+	*baseGroup
```
[the remark this answers is not in the document]
Comment (bogdan-rosianu) [reply api/groups/addressGroup.go:48]: done

File: api/groups/addressGroup.go [hunk 18]
```diff
@@ -0,0 +1,707 @@
...
+		},
+	}
+	ag.endpoints = endpoints
+
+	return ag, nil
+}
+
+// addressGroup returns a response containing information about the account correlated with provided address
+func (ag *addressGroup) getAccount(c *gin.Context) {
+	addr := c.Param("address")
```
[the remark this answers is not in the document]
Comment (bogdan-rosianu) [reply api/groups/addressGroup.go:150]: Lines `L150-L162` are just fetching the "address" parameter and respond with an error otherwise. Also, I cannot extract it as there are specific errors used (ErrGetBalance, ErrGetNonce, and so on)

File: api/middleware/endpointThrottler.go [hunk 19]
```diff
@@ -68,3 +68,46 @@ func CreateEndpointThrottler(throttlerName string) gin.HandlerFunc {
 		c.Next()
 	}
 }
+
+// CreateEndpointThrottler will create a middleware-type of handler to be used in conjunction with special
```
[the remark this answers is not in the document]
Comment (bogdan-rosianu) [reply api/middleware/endpointThrottler.go:72]: done. in the next PR, I'll delete the other function

**Review by BeniaminDrasovean — APPROVED** [review]

File: api/groups/addressGroup.go [hunk 20]
```diff
@@ -68,14 +68,9 @@ type esdtNFTTokenData struct {
...
-func NewAddressGroup(facadeHandler interface{}) (*addressGroup, error) {
-	if facadeHandler == nil {
-		return nil, errors.ErrNilFacadeHandler
-	}
-
-	facade, ok := facadeHandler.(addressFacadeHandler)
-	if !ok {
-		return nil, fmt.Errorf("%w for address group", errors.ErrFacadeWrongTypeAssertion)
+func NewAddressGroup(facade addressFacadeHandler) (*addressGroup, error) {
+	if facade == nil {
```
Comment (iulianpascalau) [inline api/groups/addressGroup.go:72]: check if nil for all tests like this one

File: api/groups/blockGroup.go (hunk 2 above)
Comment (iulianpascalau) [reply api/groups/blockGroup.go:156]: great

**Review by iulianpascalau — APPROVED** [review]

File: api/groups/addressGroup.go (hunk 20 above)
Comment (bogdan-rosianu) [reply api/groups/addressGroup.go]: done

**Review by iulianpascalau — APPROVED** [review]

**Review by BeniaminDrasovean — APPROVED** [review]
